File: daily/Binary Search/1201-Ugly-number-iii.py
class Solution:
    def nthUglyNumber(self, n: int, a: int, b: int, c: int) -> int:
        ab = self.lcm(a, b)
        ac = self.lcm(a, c)
        bc = self.lcm(b, c)
        abc = self.lcm(ab, c)

        left, right = 1, min(a, b, c) * n

        while left < right:
            mid = left + (right - left) // 2
            if self.count(mid, a, b, c, ab, ac, bc, abc) >= n:
                right = mid
            else:
                left = mid + 1

        return left

    def count(self, x: int, a: int, b: int, c: int,
              ab: int, ac: int, bc: int, abc: int) -> int:
        return (
            x // a + x // b + x // c
            - x // ab - x // ac - x // bc
            + x // abc
        )

    def gcd(self, a: int, b: int) -> int:
        while b != 0:
            a, b = b, a % b
        return a

    def lcm(self, a: int, b: int) -> int:
        return a // self.gcd(a, b) * b

File: daily/Binary Search/1201-Ugly-number-iii.go
package daily

func nthUglyNumber(n int, a int, b int, c int) int {
	ab := lcm(a, b)
	ac := lcm(a, c)
	bc := lcm(b, c)
	abc := lcm(ab, c)

	lo, hi := 1, min(a, min(b, c))*n
	for lo < hi {
		mid := lo + (hi-lo)/2
		if countUglyNumbers(mid, a, b, c, ab, ac, bc, abc) >= n {
			hi = mid
		} else {
			lo = mid + 1
		}
	}
	return lo
}

func countUglyNumbers(x, a, b, c, ab, ac, bc, abc int) int {
	return x/a + x/b + x/c -
		x/ab - x/ac - x/bc +
		x/abc
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func lcm(a, b int) int {
	return a / gcd(a, b) * b
}
